include/openthread/thread.h

```diff
@@ -153,6 +153,35 @@ typedef struct otIpCounters
     uint32_t mRxFailure; ///< The number of IPv6 packets failed to receive.
 } otIpCounters;
 
+/**
+ * This structure represents the Thread MLE counters.
+ *
+ */
+typedef struct otMleCounters
+{
+    uint16_t mDisabledRole;                  ///< Number of times device entered OT_DEVICE_ROLE_DISABLED role.
+    uint16_t mDetachedRole;                  ///< Number of times device entered OT_DEVICE_ROLE_DETACHED role.
+    uint16_t mChildRole;                     ///< Number of times device entered OT_DEVICE_ROLE_CHILD role.
+    uint16_t mRouterRole;                    ///< Number of times device entered OT_DEVICE_ROLE_ROUTER role.
+    uint16_t mLeaderRole;                    ///< Number of times device entered OT_DEVICE_ROLE_LEADER role.
+    uint16_t mAttachAttempts;                ///< Number of attach attempts while device was detached.
+    uint16_t mParitionIdChanges;             ///< Number of changes to partition ID.
+    uint16_t mBetterPartitionAttachAttempts; ///< Number of attempts to attach to a better partition.
+
+    /**
+     * Number of times device changed its parents.
+     *
+     * Support for this counter requires the feature option OPENTHREAD_CONFIG_INFORM_PREVIOUS_PARENT_ON_REATTACH to be
```

Comment:
Hi @abtink, sorry to comment on an old PR but I have a question on this change.  Can you help me understand why we require  `OPENTHREAD_CONFIG_INFORM_PREVIOUS_PARENT_ON_REATTACH`? I understand the rest of the code is setup to support this but I'm wondering why the `mParentChanges` counter doesn't track parent changes even if we choose not to enable `OPENTHREAD_CONFIG_INFORM_PREVIOUS_PARENT_ON_REATTACH`.  Thanks

Reply:
@romacdon , I double-checked the code, I think the main reason is that the logic and variables (`mPreviousParentRloc`) that remember the old parent are only available and used under the `INFORM_PREVIOUS_PARENT_ON_REATTACH` feature. Otherwise this is not tracked. 

I think it is possible (and should straight-forward) to re-shuffle the code and have the `mPreviousParentRloc` be included under all configs.

--------------

On a separate point related to `INFORM_PREVIOUS_PARNT`, I highly recommend enabling it (specially on SEDs).  This feature came out of issues/needs we observed in actual deployment (during Nest Secure development and alpha/beta tests). 
- Without this, if a child switches its parent, the old parent can still consider it as its child till the child entry times out in the old parent's child table (which can be minutes). 
- The old parent would then continue to respond to address queries for the sleepy child addresses.
- Messages intended of the child are  routed to the wrong parent (and dropped) and the old parent continues to collect and queue messages for the child (waiting for the child to poll). 

With `INFORM_PREVIOUS_PARENT` when the child attaches to a new parent, it sends an empty message to its old parent which triggers the parent to kick the child out of its child table quickly.

Reply:
Thanks for the follow up @abtink.  Couldn't we just move the counter, `mParentChanges`, to be included under all configs and leave `mPreviousParentRloc` under the control of the `INFORM_PREVIOUS_PARENT_ON_REATTACH`?

Reply:
The `mParentChanges` counter is using the `mPreviousParentRloc` to determine when/if parent was switched. 

The counter logic can be changed to have its own (potentially different) mechanism (for tracking previous parent and determining a switch) but I guess easier way is to reuse tracking of `mPreviousParentRloc` (maybe just moving it out of `INFORM_PREVIOUS_PARENT` condition).

Reply:
I see @abtink . Thanks for the explanation.  Is this something you can do?

Reply:
I can help take care of this later.

But let me again suggest to not disable `INFORM_PREVIOUS_PARENT`. Without it SEDs may fail to receive any traffic for sometime (child timeout interval which can be several minutes) after a parent switch (the situation I explained above can unfortunately happen more often than expected).

Reply:
@abtink thanks again

Reply:
Just submitted 
- https://github.com/openthread/openthread/pull/7134

Reply:
Thanks @abtink - I just reviewed #7134 and approved it!